### ISSUE TITLE
Will using this method result in inconsistent output results?

### ISSUE THREAD

**Issue by niyunsheng** [FasterDecoding/Medusa#56]
Is the results using Medusa different with the result using greedy decoding ?

**Comment by leeyeehoo** [comment]
In most cases, no. But there is a very tricky bug that we don't know is due to float16 or something else that if it will output inconsistent results after error accumulation (very rare, but it exists).

**Comment by niyunsheng** [comment]
Thanks for you reply. According to my understanding, if the calculation error is ignored (for example, using fp32), the output result should be the same as greedy decoding.

**Comment by Junyoungpark** [comment]
Hi, 

I love your work and thank you for sharing the code. I was wondering whether Medusa gives the same results to the original model (e.g., Vicuna) in sampling mode. Please let me know!

**Comment by leeyeehoo** [comment]
Hi Junyoung, we will release a new version soon and it should contain the sampling method consistent with the original model's distribution. If you want to implement it yourself, it is super easy. 

Let's say you have a medusa head's prediction x_{hat} (top-k) from medusa logits p. When you perform the verification using the model's output p, sample x from p and if x matches x_{hat} then accept x_{hat} else reject. The distribution should be the same as the original and you will get a slightly low acceleration rate.

**Comment by Junyoungpark** [comment]
Thank you for the reply. I will stay tuned for the new release!

**Comment by leeyeehoo** [comment]
If you want to check it out first, you can switch to the v1.0 branch and it should be compatible with the previous version (almost done, only the new weight hasn't been released yet.). Let me know if you encounter any issues :)

**Comment by ctlllll** [comment]
This thread seems to be quiet for a while. Let me close it for now, and feel free to reopen it :)

**Comment by zhoujian-z** [comment]
@leeyeehoo the medusa no penalty when process logits compare with base model, may be also the reason?
